Vertex AI Gemini: fix streaming model: introduce Executor to make it async

douglas-DS opened the pull request:
## Issue
Fixes #3288

## Change
- Added an `Executor` call in the `generate` method to improve asynchronous task handling by using a thread executor.
- Update logger target class for accurate logging context.
- Remove unnecessary null check for `allowedFunctionNames` property.
- Fixed empty arguments in function call.

## General checklist
- [ ] There are no breaking changes
- [ ] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

douglas-DS committed:
feat: introduce virtual thread executor for improved concurrency in streaming operations

kpavlov commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -52,7 +55,7 @@
 public class VertexAiGeminiStreamingChatModel implements StreamingChatModel, Closeable {
 
-    private static final Logger logger = LoggerFactory.getLogger(VertexAiGeminiChatModel.class);
+    private static final Logger logger = LoggerFactory.getLogger(VertexAiGeminiStreamingChatModel.class);
```
👍🏻

dliubarskyi replied:
👍🏻

kpavlov reviewed (commented):
@douglas-DS , are you sure there is no native calls or synchronized blocks in the underlaying code?

Please refer to 
- https://quarkus.io/guides/virtual-threads
- https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything
- https://quarkus.io/blog/virtual-threads-2/

Let's add thread pinning detection (`-Djdk.tracePinnedThreads`) to the java 21 maven profile: 
https://quarkus.io/guides/virtual-threads#detect-pinned-thread-in-tests

douglas-DS commented:
> @douglas-DS , are you sure there is no native calls or synchronized blocks in the underlaying code?
> 
> Please refer to
> 
>     * https://quarkus.io/guides/virtual-threads
> 
>     * https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything
> 
>     * https://quarkus.io/blog/virtual-threads-2/
> 
> 
> Let's add thread pinning detection (`-Djdk.tracePinnedThreads`) to the java 21 maven profile: https://quarkus.io/guides/virtual-threads#detect-pinned-thread-in-tests

What do you think would be a better approach here?

I’m wondering whether it would be better to pass the executor to the builder class.

kpavlov commented:
Passing the executor to builder class is very good approach. Even making it a general pattern by introducing abstract builder and inheriting from it. But I am just worrying about virtual thread safety of the underlaying vertex api.
Yes, making it customizable via builder would not break things for everyone, but will allow to enable virtual threads in your application if you want.

kpavlov commented:
And let's add jvm flag to report thread pinning. It might help to detect such issues in other places

douglas-DS committed:
feat: add default executor and allow custom executor configuration

douglas-DS committed:
feat: handle empty tool execution arguments gracefully and add test coverage

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -52,7 +56,9 @@
-    private static final Logger logger = LoggerFactory.getLogger(VertexAiGeminiChatModel.class);
+    private static final Logger logger = LoggerFactory.getLogger(VertexAiGeminiStreamingChatModel.class);
+
+    private static final Executor DEFAULT_EXECUTOR = Executors.newCachedThreadPool();
```
Can we create default executor lazily, only if it is not provided by the user?
Also it it probably better to have a dedicated executor per each `VertexAiGeminiStreamingChatModel` instance.

dliubarskyi replied:
We also need to make sure this executor shuts down by itself if not used for some (short) time, otherwise it will prevent the user application to shutdown.
See an example here: https://github.com/langchain4j/langchain4j/blob/536705ecff46d9749f3d8711da37d0ffd1131b6c/langchain4j-core/src/main/java/dev/langchain4j/rag/DefaultRetrievalAugmentor.java#L128

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -168,6 +174,7 @@ public VertexAiGeminiStreamingChatModel(VertexAiGeminiStreamingChatModelBuilder
         this.logRequests = getOrDefault(builder.logRequests, false);
         this.logResponses = getOrDefault(builder.logResponses, false);
         this.listeners = copy(builder.listeners);
+        this.executor = Objects.requireNonNullElse(builder.executor, DEFAULT_EXECUTOR);
```
```suggestion
        this.executor = getOrElse(builder.executor, () -> createDefaultExecutor());
```

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -371,105 +372,106 @@ private void generate(
-            handler.onError(exception);
-        }
+            try {
+                model.generateContentStream(instructionAndContent.contents).stream().forEach(partialResponse -> {
```
Maybe let's execute code in the executor starting from this line? All the request preparation is better to do in the caller thread (in case there will be some validation exceptions thrown)

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/FunctionCallHelper.java:
```diff
@@ -25,7 +25,9 @@ static FunctionCall fromToolExecutionRequest(ToolExecutionRequest toolExecutionR
         try {
-            JsonFormat.parser().merge(toolExecutionRequest.arguments(), structBuilder);
+            String toolArguments = toolExecutionRequest.arguments();
+            String arguments = toolArguments.isBlank() ? "{}" : toolArguments;
```
```suggestion
            String arguments = isNullOrBlank(toolArguments) ? "{}" : toolArguments;
```

dliubarskyi reviewed (commented):
@douglas-DS thank you!

douglas-DS committed:
fix: handle null tool execution arguments in FunctionCallHelper

douglas-DS committed:
feat: configure surefire plugin to support custom system properties and JVM arguments

douglas-DS committed:
fix: use utility method to check null or blank tool execution arguments

douglas-DS committed:
fix: allow custom executor in VertexAiGeminiStreamingChatModel

Refactored executor handling to enable custom executors and replaced the default executor implementation with a thread pool executor for better control. Updated relevant constructors and tests.

douglas-DS committed:
refactor: call executor after build request logic

douglas-DS commented:
@dliubarskyi I've made all the requested changes. Could you please review it again?

kpavlov committed:
Merge branch 'main' into fix/virtual-thread-executor

kpavlov commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -321,6 +324,15 @@ public VertexAiGeminiStreamingChatModel(GenerativeModel generativeModel, Generat
+
+    private static ExecutorService createDefaultExecutor() {
+        return new ThreadPoolExecutor(
+            0, Integer.MAX_VALUE,
```
max pool size should not be Integer.MAX_VALUE

dliubarskyi replied:
Why not? `Executors.newCachedThreadPool()` does exactly that

kpavlov replied:
Unconstrained thread pool is a typically is not good for production as it can may consume too many platform threads. The JDK implementation is optimistic, although it's difficult to come up with exact value.
I am fine to leave it unconstrained, since it is possible to tune for the specific application.

douglas-DS replied:
To be honest, I don’t see any benefit in having granular control over the default executor, since we already have the option to pass a specific executor implementation through the builder if needed.

douglas-DS commented:
@dliubarskyi @kpavlov  Is this ready for merge?

kpavlov committed:
Merge branch 'main' into fix/virtual-thread-executor

kpavlov commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java:
```diff
@@ -482,6 +497,12 @@ public void close() {
         }
+
+        if (this.executor != null) {
+            if (executor instanceof ExecutorService) {
```
can be simplified: if executor is null - it is not instance of ExecutorService. `if (this.executor != null)` is not needed

kpavlov commented on langchain4j-vertex-ai-gemini/src/test/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModelIT.java:
```diff
@@ -165,7 +165,7 @@ void should_allow_custom_generativeModel_and_generationConfig() {
         GenerationConfig generationConfig = GenerationConfig.getDefaultInstance();
 
-        StreamingChatModel model = new VertexAiGeminiStreamingChatModel(generativeModel, generationConfig);
+        StreamingChatModel model = new VertexAiGeminiStreamingChatModel(generativeModel, generationConfig, null);
```
Shouldn't we create overriding method without the `executor` parameter for backwards compatibility ?

douglas-DS replied:
I agree! I’ve brought the previous constructor back. Now we have both overloaded constructors, one with it and one without it.

douglas-DS committed:
refactor: simplify executor shutdown logic in VertexAiGeminiStreamingChatModel

douglas-DS committed:
refactor: add overload constructor for executor parameter

- Introduced a new constructor without `executor` parameter.
- Updated the test case to use the simplified constructor.

douglas-DS committed:
Merge branch 'main' into fix/virtual-thread-executor

# Conflicts:
#	langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiStreamingChatModel.java

douglas-DS commented:
@kpavlov  @dliubarskyi Could you review this again?

dliubarskyi committed:
Merge branch 'main' into fix/virtual-thread-executor

dliubarskyi reviewed (approved):
@douglas-DS thanks a lot!
